alg/gdal_homography.cpp

```diff
@@ -0,0 +1,619 @@
+        max_line = std::max(max_line, pasGCPList[i].dfGCPLine);
+        min_geox = std::min(min_geox, pasGCPList[i].dfGCPX);
+        max_geox = std::max(max_geox, pasGCPList[i].dfGCPX);
+        min_geoy = std::min(min_geoy, pasGCPList[i].dfGCPY);
+        max_geoy = std::max(max_geoy, pasGCPList[i].dfGCPY);
+    }
+
+    double EPS = 1.0e-12;
+
+    if (std::abs(max_pixel - min_pixel) < EPS ||
+        std::abs(max_line - min_line) < EPS ||
+        std::abs(max_geox - min_geox) < EPS ||
+        std::abs(max_geoy - min_geoy) < EPS)
+    {
+        return FALSE;  // degenerate in at least one dimension.
+    }
+
+    double pl_normalize[9], geo_normalize[9];
+
+    pl_normalize[0] = -min_pixel / (max_pixel - min_pixel);
+    pl_normalize[1] = 1.0 / (max_pixel - min_pixel);
+    pl_normalize[2] = 0.0;
+    pl_normalize[3] = -min_line / (max_line - min_line);
+    pl_normalize[4] = 0.0;
+    pl_normalize[5] = 1.0 / (max_line - min_line);
+    pl_normalize[6] = 1.0;
+    pl_normalize[7] = 0.0;
+    pl_normalize[8] = 0.0;
+
+    geo_normalize[0] = -min_geox / (max_geox - min_geox);
+    geo_normalize[1] = 1.0 / (max_geox - min_geox);
+    geo_normalize[2] = 0.0;
+    geo_normalize[3] = -min_geoy / (max_geoy - min_geoy);
+    geo_normalize[4] = 0.0;
+    geo_normalize[5] = 1.0 / (max_geoy - min_geoy);
+    geo_normalize[6] = 1.0;
+    geo_normalize[7] = 0.0;
+    geo_normalize[8] = 0.0;
+
+    double inv_geo_normalize[9] = {0.0};
+    if (!GDALInvHomography(geo_normalize, inv_geo_normalize))
+    {
+        return FALSE;
+    }
+
+    GDALMatrix LtL(9, 9);
+    GDALMatrix rhs(9, 1);
+    rhs(6, 0) = 1;
+    LtL(6, 6) = 1;
+
+    for (int i = 0; i < nGCPCount; ++i)
+    {
+        double pixel, line, geox, geoy;
+
+        GDALApplyHomography(pl_normalize, pasGCPList[i].dfGCPPixel,
+                            pasGCPList[i].dfGCPLine, &pixel, &line);
+        GDALApplyHomography(geo_normalize, pasGCPList[i].dfGCPX,
+                            pasGCPList[i].dfGCPY, &geox, &geoy);
+
+        double Lx[] = {1, pixel, line,          0,           0,
+                       0, -geox, -geox * pixel, -geox * line};
+        double Ly[] = {0,           0, 0, 1, pixel, line, -geoy, -geoy * pixel,
+                       -geoy * line};
+        int j, k;
+        // Populate the lower triangle of symmetric LtL matrix
+        for (j = 0; j < 9; j++)
+        {
+            for (k = j; k < 9; k++)
+            {
+                LtL(j, k) += Lx[j] * Lx[k] + Ly[j] * Ly[k];
+            }
+        }
+    }
+    // Populate the upper triangle of symmetric LtL matrix
+    for (int j = 0; j < 9; j++)
+    {
+        for (int k = 0; k < j; k++)
+        {
+            LtL(j, k) = LtL(k, j);
+        }
+    }
+
+    GDALMatrix h_normalized(9, 1);
+    if (!GDALLinearSystemSolve(LtL, rhs, h_normalized))
+    {
+        return FALSE;
+    }
+
+    // "Hour-glass" like shape of GCPs. Cf https://github.com/OSGeo/gdal/issues/11618
```

Comment:
I don't understand that comment. Hour-glass like shape is one possibility, not necesarily all the cases